### ISSUE TITLE
Incorporating FastSurfer to sMRIPrep as surface reconstruction option

### ISSUE THREAD

**Issue by pcamach2** [nipreps/smriprep#278]
**Is your feature request related to a problem? Please describe.**
No

**Describe the solution you'd like**
Addition of FastSurfer workflows as an alternative to FreeSurfer.

**Describe alternatives you've considered**
I have considered just running FastSurfer before running sMRIPrep or fMRIPrep and using the outputs as an existing Freesurfer output directory. This may work as discussed in #273 by @effigies.

**Additional context**
I brought this up in the Nipype discourse group and @oesteban suggested moving the discussion here on GitHub.
I have a working [Dockerfile](https://github.com/pcamach2/smriprep/blob/master/Dockerfile) that includes FastSurfer and sMRIPrep, which I have used to run FastSurfer on my workstation.

Currently, building the Docker image requires having downloaded a local copy of Convert3D since my build machine was throwing errors for ca-certificates when trying to curl the tar.gz from SourceForge as part of the build. Testing this image for FastSurfer - once built as smriprep:fastsurfer_dev - can be performed with the following command:

```
docker run --gpus all --entrypoint /bin/bash -v /path/to/bids:/data \
-v /path/to/bids/derivatives/fastsurfer:/output \
-v /path/to/freesurfer/license:/fs60/license --rm --user `id -u` \
smriprep:fastsurfer_dev /opt/FastSurfer/run_fastsurfer.sh --fs_license /fs60/license \
--t1 /data/sourcedata/sub-SUBJECT/ses-SESSION/anat/sub-SUBJECT_ses-SESSION_T1w.nii.gz \
--sid sub-SUBJECT --sd /output
```

I am building a Singularity image to test on our cluster and will test both tools separately before scripting this into the relevant nipype workflows.  

I appreciate any help and/or feedback!

**Comment by pcamach2** [comment]
- [x] Complete runs of FastSurfer GPU in both the Docker and Singularity versions of this container 
- [ ] Complete runs of sMRIPrep in both the Docker and Singularity versions of this container. 
- [ ] Build into Nipype workflow

**Comment by pcamach2** [comment]
Any help with integrating this to Nipype would be very much appreciated!

**Comment by effigies** [comment]
Hi @pcamach2, thanks for this!

To create a nipype interface for FastSurfer, we're going to need to know a few things:

1) What does a command-line look like? What inputs does it have, what options does it have?
2) Is there a notion of resuming, as there is in FreeSurfer? Do we need to include machinery to detect partially complete runs?

**Comment by pcamach2** [comment]
Hi @effigies, no problem!

Here is what I have so far:

>     1. What does a command-line look like? What inputs does it have, what options does it have?

Per the [basic FastSurfer container run instructions](https://github.com/Deep-MI/FastSurfer/tree/master/Docker#example-1-build-gpu-fastsurfer-container-default), the barebones docker run command is as follows for their Docker image:

```
docker run --gpus all -v /home/user/my_mri_data:/data \
                      -v /home/user/my_fastsurfer_analysis:/output \
                      -v /home/user/my_fs_license_dir:/fs60 \
                      --rm --user XXXX fastsurfer:gpu \
                      --fs_license /fs60/.license \
                      --t1 /data/subject2/orig.mgz \
                      --sid subject2 --sd /output \
                      --parallel
```

From the [hybrid container that I built](https://github.com/pcamach2/smriprep/blob/master/Dockerfile), my tests ran with the following (with all <label> replaced with actual labels from the test dataset):

**For the Singularity container (running from the project directory):**

```
singularity exec --nv -B ./bids:/data,./bids/derivatives/fastsurfer:/output,/path/to/freesurferlicense/license.txt:/fs60/license \
/path/to/smriprep-fastsurfer_dev.sif /opt/FastSurfer/run_fastsurfer.sh --fs_license /fs60/license \
--t1 /data/sub-<label>/ses-<label>/anat/sub-<label>_ses-<label>_T1w.nii.gz --sid sub-<label> --sd /output
```

**From the shell in the container:**
```
/opt/FastSurfer/run_fastsurfer.sh --fs_license /fs60/license \
--t1 /data/sub-<label>/ses-<label>/anat/sub-<label>_ses-<label>_acq-<label>_T1w.nii.gz \
--sid sub-<label> --sd /output

```
*Note that the FreeSurfer license is mounted to /fs60/license - this is just their convention for mounting.*
*This project was not set up for the new BIDS specification, so these "raw" BIDS NIFTIs should be in project/bids/sourcedata now.*


The full list of arguments is explained as follows:

[Required arguments](https://github.com/Deep-MI/FastSurfer#required-arguments):

>  --sd: Output directory $SUBJECTS_DIR (equivalent to FreeSurfer setup --> $SUBJECTS_DIR/sid/mri; $SUBJECTS_DIR/sid/surf ... will be created).
    --sid: Subject ID for directory inside $SUBJECTS_DIR to be created ($SUBJECTS_DIR/sid/...)
    --t1: T1 full head input (not bias corrected, global path). The network was trained with conformed images (UCHAR, 256x256x256, 1 mm voxels and standard slice orientation). These specifications are checked in the eval.py script and the image is automatically conformed if it does not comply.
    --fs_license: Path to FreeSurfer license key file. Register (for free) at https://surfer.nmr.mgh.harvard.edu/registration.html to obtain it if you do not have FreeSurfer installed so far. Strictly necessary if you use Docker, optional for local install (your local FreeSurfer license will automatically be used)

[Optional arguments](https://github.com/Deep-MI/FastSurfer#network-specific-arguments-optional)   
 
>   Network specific arguments (optional):
    --seg: Global path with filename of segmentation (where and under which name to store it). Default location: $SUBJECTS_DIR/$sid/mri/aparc.DKTatlas+aseg.deep.mgz
    --weights_sag: Pretrained weights of sagittal network. Default: ../checkpoints/Sagittal_Weights_FastSurferCNN/ckpts/Epoch_30_training_state.pkl
    --weights_ax: Pretrained weights of axial network. Default: ../checkpoints/Axial_Weights_FastSurferCNN/ckpts/Epoch_30_training_state.pkl
    --weights_cor: Pretrained weights of coronal network. Default: ../checkpoints/Coronal_Weights_FastSurferCNN/ckpts/Epoch_30_training_state.pkl
    --seg_log: Name and location for the log-file for the segmentation (FastSurferCNN). Default: $SUBJECTS_DIR/$sid/scripts/deep-seg.log
    --clean_seg: Flag to clean up FastSurferCNN segmentation
    --run_viewagg_on: Define where the view aggregation should be run on. By default, the program checks if you have enough memory to run the view aggregation on the gpu. The total memory is considered for this decision. If this fails, or you actively overwrote the check with setting "--run_viewagg_on cpu", view agg is run on the cpu. Equivalently, if you define "--run_viewagg_on gpu", view agg will be run on the gpu (no memory check will be done).
    --no_cuda: Flag to disable CUDA usage in FastSurferCNN (no GPU usage, inference on CPU)
    --batch: Batch size for inference. Default: 16. Lower this to reduce memory requirement
    --order: Order of interpolation for mri_convert T1 before segmentation (0=nearest, 1=linear(default), 2=quadratic, 3=cubic)

> Surface pipeline arguments (optional):
    --fstess: Use mri_tesselate instead of marching cube (default) for surface creation
    --fsqsphere: Use FreeSurfer default instead of novel spectral spherical projection for qsphere
    --fsaparc: Use FS aparc segmentations in addition to DL prediction (slower in this case and usually the mapped ones from the DL prediction are fine)
    --surfreg: Create Surface-Atlas (sphere.reg) registration with FreeSurfer (for cross-subject correspondence or other mappings)
    --parallel: Run both hemispheres in parallel
    --threads: Set openMP and ITK threads to

> Other:
    --py: which python version to use. Default: python3.6
    --seg_only: only run FastSurferCNN (generate segmentation, do not run the surface pipeline)
    --surf_only: only run the surface pipeline recon_surf. The segmentation created by FastSurferCNN must already exist in this case.


I think the most important factor that we would need to consider is whether folks are running with GPU or CPU since this could affect the image they would use. One way to handle that would be building separate versions and tag them with "GPU" or "CPU" accordingly. Otherwise, I think we can build the general GPU version by default and allow the --no_cuda argument to handle this.

>     2. Is there a notion of resuming, as there is in FreeSurfer? Do we need to include machinery to detect partially complete runs?

I haven't had to quit and resume with FastSurfer due to the shorter run time and the lack of long process stalling that can occur in recon-all, so I can't say from experience. 

The main breakpoint built in is that the segmentation via CNN can be run standalone and then followed by surface reconstruction. I have not personally taken this approach, but it might be useful if someone wants to generate a large batch of segmentations and check them before committing to surface reconstruction (although it seems more likely that they would just use standalone FastSurfer in this use case).

I would lean towards using a fresh run if all output files are not detected; this also makes it easier to assess quality with their adapted [QATools](https://github.com/Deep-MI/qatools-python), which NaNs out when the recon-all log indicates restarts.

For reference, here are the example outputs from a bids/derivatives/fastsurfer/sub-<label> directory:

```
./bids/derivatives/fastsurfer/sub-<label>/label:
aparc.annot.mapped.ctab
lh.aparc.mapped.annot
lh.aparc.mapped.prefix.annot
lh.cortex.label
rh.aparc.mapped.annot
rh.aparc.mapped.prefix.annot
rh.cortex.label

./bids/derivatives/fastsurfer/sub-<label>/mri:
aparc+aseg.orig.mgz
aparc.DKTatlas+aseg.deep.mgz
aparc.DKTatlas+aseg.deep.withCC.mgz
aparc.mapped+aseg.mgz
aseg.auto.mgz
aseg.auto_noCCseg.mgz
aseg.mgz
aseg.presurf.hypos.mgz
aseg.presurf.mgz
brain.finalsurfs.mgz
brainmask.mgz
brain.mgz
filled.mgz
filled-pretess127.mgz
filled-pretess255.mgz
lh.ribbon.mgz
mask.mgz
mri_nu_correct.mni.log
mri_nu_correct.mni.log.bak
norm.mgz
nu.mgz
orig
orig.mgz
orig_nu.mgz
rawavg.mgz
rh.ribbon.mgz
ribbon.mgz
segment.dat
T1.mgz
tmp
transforms
wm.asegedit.mgz
wm.mgz
wmparc.mapped.mgz
wm.seg.mgz

./bids/derivatives/fastsurfer/sub-<label>/scripts:
build-stamp.txt
deep-seg.log
lastcall.build-stamp.txt
lh.processing.cmdf
patchdir.txt
pctsurfcon.log
pctsurfcon.log.old
ponscc.cut.log
recon-all.cmd
recon-all.done
recon-all.env
recon-all.env.bak
recon-all.local-copy
recon-all.log
recon-all-status.log
recon-surf.done
recon-surf.log
recon-surf_times.yaml
rh.processing.cmdf

./bids/derivatives/fastsurfer/sub-<label>/stats:
aseg.presurf.hypos.stats
aseg.stats
lh.aparc.mapped.stats
lh.curv.stats
lh.w-g.pct.stats
rh.aparc.mapped.stats
rh.curv.stats
rh.w-g.pct.stats
wmparc.mapped.stats

./bids/derivatives/fastsurfer/sub-<label>/surf:
lh.area
lh.area.mid
lh.area.pial
lh.curv
lh.curv.pial
lh.defect_borders
lh.defect_chull
lh.defect_labels
lh.inflated
lh.inflated.H
lh.inflated.K
lh.inflated.nofix
lh.orig
lh.orig.nofix
lh.pial
lh.qsphere.nofix
lh.smoothwm
lh.smoothwm.BE.crv
lh.smoothwm.C.crv
lh.smoothwm.FI.crv
lh.smoothwm.H.crv
lh.smoothwm.K1.crv
lh.smoothwm.K2.crv
lh.smoothwm.K.crv
lh.smoothwm.nofix
lh.smoothwm.S.crv
lh.sulc
lh.thickness
lh.volume
lh.w-g.pct.mgh
lh.white
lh.white.H
lh.white.K
lh.white.preaparc
lh.white.preaparc.H
lh.white.preaparc.K
rh.area
rh.area.mid
rh.area.pial
rh.curv
rh.curv.pial
rh.defect_borders
rh.defect_chull
rh.defect_labels
rh.inflated
rh.inflated.H
rh.inflated.K
rh.inflated.nofix
rh.orig
rh.orig.nofix
rh.pial
rh.qsphere.nofix
rh.smoothwm
rh.smoothwm.BE.crv
rh.smoothwm.C.crv
rh.smoothwm.FI.crv
rh.smoothwm.H.crv
rh.smoothwm.K1.crv
rh.smoothwm.K2.crv
rh.smoothwm.K.crv
rh.smoothwm.nofix
rh.smoothwm.S.crv
rh.sulc
rh.thickness
rh.volume
rh.w-g.pct.mgh
rh.white
rh.white.H
rh.white.K
rh.white.preaparc
rh.white.preaparc.H
rh.white.preaparc.K

./bids/derivatives/fastsurfer/sub-<label>/tmp:

./bids/derivatives/fastsurfer/sub-<label>/touch:
asegmerge.touch
cortical_ribbon.touch
fill.touch
inorm2.touch
lh.curvstats.touch
lh.inflate1.touch
lh.inflate2.touch
lh.inflate.H.K.touch
lh.smoothwm2.touch
lh.surfvolume.touch
lh.topofix.touch
lh.white.H.K.touch
relabelhypos.touch
rh.curvstats.touch
rh.inflate1.touch
rh.inflate2.touch
rh.inflate.H.K.touch
rh.smoothwm2.touch
rh.surfvolume.touch
rh.topofix.touch
rh.white.H.K.touch
rusage.mris_fix_topology.lh.dat
rusage.mris_fix_topology.rh.dat
rusage.mris_inflate.lh.dat
rusage.mris_inflate.rh.dat
segstats.touch
wmsegment.touch

./bids/derivatives/fastsurfer/sub-<label>/trash:
```

Thank you!

**Comment by pcamach2** [comment]
@effigies, the fix of running `touch` on the `mri/aseg.auto_noCCseg.label_intensities.txt` file (https://github.com/Deep-MI/FastSurfer/issues/21#issuecomment-1103950005) resolves the error and allows sMRIPrep to complete.

I added a short python function to my fork of smriprep/smriprep/utils/misc.py (https://github.com/pcamach2/smriprep/blob/1377f4c629542ba8d3ab11f90b9c593ffa29b6cc/smriprep/utils/misc.py) that checks for FastSurfer files in the --fs-subjects-dir (there are .mgz files with 'deep' in the subject's mri directory, which indicate that they were generated using the deep-learning method used in FastSurfer). If detected, this function runs the `touch` fix and sets a `fastsurfer_bool` variable to 'True'.

I think the next step would be to use the `fastsurfer_bool` variable in anatomical.py to run minimal Freesurfer commands, but I would appreciate any advice on where best to call the `check_fastsurfer` function and whether I should create a separate FastSurfer-specific workflow in surfaces.py. 

Thank you!

**Comment by pcamach2** [comment]
I removed the `fastsurfer_bool` variable and added the `check_fastsurfer` function to anatomical.py (https://github.com/pcamach2/smriprep/blob/pcamach2-fastsurfer-patch/smriprep/workflows/anatomical.py), which successfully builds the anat workflow and touches the label_intensities file. This runs as expected, so the failure to complete when using FastSurfer outputs as --fs-subjects-dir appears to be resolved.

The next steps I can think of are:
- [ ] setting up the alternative workflow with minimal Freesurfer commands
- [ ] building a nipype FastSurfer wrapper
- [ ] adjusting documentation and boilerplate to reflect FastSurfer being used
- [ ] adding a FastSurfer argument to the smriprep cli

**Comment by pcamach2** [comment]
@effigies, I am trying to connect the FastSurfer interface that I wrote and tested (https://github.com/pcamach2/smriprep/blob/pcamach2-fastsurfer-patch-dev/smriprep/interfaces/fastsurfer.py) into the overall sMRIPrep workflow.

I added the following argument to the cli (https://github.com/pcamach2/smriprep/blob/pcamach2-fastsurfer-patch-dev/smriprep/cli/run.py)
```
    g_surfs_xor.add_argument(
        "--fastsurfer-recon",
        action="store_true",
        dest="run_fastsurfer",
        help="enable FastSurfer surface preprocessing.",
    )
```

_I am wondering where best to split workflow behavior (in anatomical.py?) based on the above argument_

Would it be better to make separate surfaces.py versions for FastSurfer and FreeSurfer or just different workflows within a unified surfaces.py? _Note that if a dataset includes FLAIR or T2w, FreeSurfer will still want to process them automatically even if there is FastSurfer output (this adds on a few hours)._ 

Thank you again for all your help and advice!

**Comment by effigies** [comment]
@pcamach2 Could you go ahead and open a draft PR with your changes? I think it will make it easier to look through this. Unfortunately I'm out of time today, but I'll put you on my todo list for tomorrow.
